File: DSA/Linkedlist/Code/DoublyLinkedList.cpp
#include <iostream>
using namespace std;

// Doubly Linked List in C++
struct Node {
    int data;
    Node* next;
    Node* prev;
    Node(int data) : data(data), next(NULL), prev(NULL) {}
};

class DoublyLinkedList {
private:
    Node* head;
    
public:
    DoublyLinkedList() : head(NULL) {}
    
    void insertAtBeginning(int data) {
        Node* newNode = new Node(data);
        if (head != NULL)
            head->prev = newNode;
        newNode->next = head;
        head = newNode;
    }
    
    void insertAtEnd(int data) {
        Node* newNode = new Node(data);
        if (head == NULL) {
            head = newNode;
            return;
        }
        
        Node* temp = head;
        while (temp->next != NULL)
            temp = temp->next;
        
        temp->next = newNode;
        newNode->prev = temp;
    }
    
    void displayForward() {
        Node* temp = head;
        while (temp != NULL) {
            cout << temp->data << " <-> ";
            temp = temp->next;
        }
        cout << "NULL\n";
    }
    
    void displayBackward() {
        if (head == NULL) return;
        Node* temp = head;
        while (temp->next != NULL)
            temp = temp->next;
        
        while (temp != NULL) {
            cout << temp->data << " <-> ";
            temp = temp->prev;
        }
        cout << "NULL\n";
    }
    
    ~DoublyLinkedList() {
        Node* temp;
        while (head != NULL) {
            temp = head;
            head = head->next;
            delete temp;
        }
    }
};

int main() {
    DoublyLinkedList list;
    
    cout << "=== Doubly Linked List ===\n\n";
    
    list.insertAtEnd(10);
    list.insertAtEnd(20);
    list.insertAtEnd(30);
    
    cout << "Forward: ";
    list.displayForward();
    
    cout << "Backward: ";
    list.displayBackward();
    
    list.insertAtBeginning(5);
    cout << "\nAfter inserting 5 at beginning:\n";
    cout << "Forward: ";
    list.displayForward();
    
    cout << "\nComplexity Analysis:\n";
    cout << "Insert at beginning: O(1)\n";
    cout << "Insert at end: O(n)\n";
    cout << "Delete: O(1) with pointer, O(n) without\n";
    cout << "Space: O(n)\n";
    
    return 0;
}

File: DSA/Linkedlist/Code/DoublyLinkedList.py
# Doubly Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
    
    def insert_at_beginning(self, data):
        new_node = Node(data)
        if self.head is not None:
            self.head.prev = new_node
        new_node.next = self.head
        self.head = new_node
    
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        temp.next = new_node
        new_node.prev = temp
    
    def display_forward(self):
        temp = self.head
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print(" <-> ".join(elements) + " <-> NULL")
    
    def display_backward(self):
        if self.head is None:
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.prev
        print(" <-> ".join(elements) + " <-> NULL")

# Test
dll = DoublyLinkedList()

print("=== Doubly Linked List ===\n")

dll.insert_at_end(10)
dll.insert_at_end(20)
dll.insert_at_end(30)

print("Forward: ", end="")
dll.display_forward()

print("Backward: ", end="")
dll.display_backward()

dll.insert_at_beginning(5)
print("\nAfter inserting 5 at beginning:")
print("Forward: ", end="")
dll.display_forward()

print("\nComplexity Analysis:")
print("Insert at beginning: O(1)")
print("Insert at end: O(n)")
print("Delete: O(1) with pointer, O(n) without")
print("Space: O(n)")
